---
title: "Hierfstat latest features"
author: "Jerome Goudet"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Hierfstat latest features}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage{utf8}{inputenc}
---

#Introduction

This vignette documents the latest developments in `hierfstat`. Refer to the other documents for an introduction to the package

#Loading data

`hierfstat` can now read `genind` objects (from package `adegenet`). Note however that only some genetic data types will be properly converted and used.  The alleles need to be encoded either as integer (up to three digits per allele), or as nucleotides (`c("a","c","g","t","A","C","G","T")`).

```{r,message=FALSE}
library(adegenet)
library(hierfstat)
```

```{r}
data(nancycats) 
is.genind(nancycats)
```
#Descriptive statistics

The function you are the most likely to want using is `basic.stats` (it calculates $H_O$, $H_S$, $F_{IS}$, $F_{ST}$ etc...). 

```{r}
bs.nc<-basic.stats(nancycats)
bs.nc
boxplot(bs.nc$perloc[,1:3]) # boxplot of Ho, Hs, Ht
```

You can also get e.g. `allele.count` and `allelic.richness`, a rarefied measure of the number of alleles at each locus and in each population. For instance, below is a boxplot of the allelic richness for the 5 first loci in the nancycats dataset

```{r}
boxplot(t(allelic.richness(nancycats)$Ar[1:5,])) #5 first loci
```

#Populations 

Population statistics are obtained through `basic.stats` or `wc` (`varcomp.glob` can also be used and will give the same result as `wc` for a one level hierarchy). For instance, $F_{IS}$ and $F_{ST}$ in the *Galba truncatula* dataset provided with `hierfstat` are obtained as:

```{r}
data(gtrunchier) 
wc(gtrunchier[,-2])
varcomp.glob(data.frame(gtrunchier[,1]),gtrunchier[,-c(1:2)])$F #same
```

Confidence intervals on these statistics can be obtained via `boot.vc`:

```{r}
boot.vc(gtrunchier[,1],gtrunchier[,-c(1:2)])$ci
```

`boot.ppfis`and `boot.ppfst` provide bootsrap confidence intervals (bootstrapping over loci) for population specific $F_{IS}$ and pairwise $F_{ST}$ respectively. 

#Genetic distances

`genet.dist` estimates one of 10 different genetic distances between populations as described mostly in Takezaki & Nei (1996)

```{r}
(Ds<-genet.dist(gtrunchier[,-2],method="Ds")) # Nei's standard genetic distances
```


#Population Principal coordinate analysis

Principal coordinate analysis can be carried out on this genetic distances:

```{r,eval=FALSE}
pcoa(as.matrix(Ds))
```

#Individual PCA

`hierfstat` has a function `indpca` carrying out Principal component analysis on individual genotypes

```{r}
x<-indpca(gtrunchier[,-2],ind.labels=gtrunchier[,2])
plot(x,col=gtrunchier[,1],cex=0.7)
```

#Sex-biased dispersal

A new function to detect sex-biased dispersal, `sexbias.test` based on [Goudet et al. (2002)](http://onlinelibrary.wiley.com/doi/10.1046/j.1365-294X.2002.01496.x/abstract) has been implemented. 
To illustrate its use, load the *Crocidura russula* data set. It consists of the genotypes and sexes of 140 shrews studied by [Favre et al. (1997)](http://rspb.royalsocietypublishing.org/content/264/1378/127). In this species, mark -recapture showed an excess of dispersal from females, an unusual pattern in mammals. This is confirmed using genetic data: 

```{r}
data("crocrussula")
aic<-AIc(crocrussula$genot)
boxplot(aic~crocrussula$sex)
tapply(aic,crocrussula$sex,mean)
sexbias.test(crocrussula$genot,crocrussula$sex)
```
#Simulating genetic data

It is now possible to simulate genetic data from a continent islands model, either at equilibrium via `sim.genot` or for a given number of generations via `sim.genot.t`.  These two functions have several arguments, allowing to look at the effect of population sizes, inbreeding, migration and mutation. the number of independant loci and number of alleles per loci can also be specified. It is also possible to simulate data from a finite island model using `sim.genot.t`, by specifying that argument `IIM` is `FALSE`.  


#Exporting to other programs

`hierfstat`can now export to files suitable for [Bayescan](http://cmpg.unibe.ch/software/BayeScan/), [plink](http://pngu.mgh.harvard.edu/~purcell/plink/) and [structure](http://pritchardlab.stanford.edu/structure.html). Functions to import from these program can be found in `adegenet`. The functions to export to these programs are `write.bayescan`, `write.ped` and `write.struct`respectively.



