<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Jerome Goudet" />

<meta name="date" content="2023-03-27" />

<title>Importing data in Hierfstat</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Importing data in Hierfstat</h1>
<h4 class="author">Jerome Goudet</h4>
<h4 class="date">2023-03-27</h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This vignette documents how to import or enter genotypic data for the
<code>hierfstat</code> package. Originally this package was written to
estimate and test hierarchical F-statistics, but was then further
developed and now include almost all features of the <a href="https://www2.unil.ch/popgen/softwares/fstat.htm">Fstat program</a>
(no longer maintained), as well as several others.</p>
</div>
<div id="format-required-by-most-functions-in-hierfstat" class="section level1">
<h1>Format required by most functions in <em>Hierfstat</em></h1>
<p>The data types that <code>hierfstat</code> can analyse are haploid or
diploid, unphased, multilocus genotypes. Note that each data set must be
made of only one ploidy level.</p>
<p>The basic data structure required by most <code>Hierfstat</code>
function is a data frame with the first column containing a population
identifier (preferably a number), and the next <span class="math inline">\(nl\)</span> columns the genotype at each of <span class="math inline">\(nl\)</span> loci.</p>
<p>In <code>hierfstat</code>, alleles are encoded as 1, 2 or 3 digits
numbers, and genotypes are encoded as numbers with the two alleles
collated (as in pasted together). Other type of data can be imported
(see below) but for the time being we focus on the primary data type.
Thus imagine that you have an individual genotyped at a microsatellite
locus with allele length <code>120</code> and <code>124</code>, the way
to encode it for <code>hierfstat</code> is either <code>120124</code>or
<code>124120</code>. If the data are SNPs, each allele at a locus could
be encoded as <code>1</code> and <code>2</code>, or you may decide to
keep the correspondence between nucleotides and alleles
(e.g. <code>1, 2, 3, 4</code> for <code>A, C, G, T</code>). Thus, if the
two alleles at a SNP locus are <code>A</code> and <code>T</code> and an
individual is heterozygote, it could be encoded as <code>14</code>or
<code>41</code>.</p>
<p>Example data sets are included in <code>hierfstat</code>. For
instance:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(hierfstat) <span class="co">#load the library</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(diploid) <span class="co"># info about this data set with ?diploid</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(diploid)</span></code></pre></div>
<pre><code>##   Pop loc-1 loc-2 loc-3 loc-4 loc-5
## 1   1    44    43    43    33    44
## 2   1    44    44    43    33    44
## 3   1    44    44    43    43    44
## 4   1    44    44    NA    33    44
## 5   1    44    44    24    34    44
## 6   1    44    44    NA    43    44</code></pre>
<p>The first individual (first row of the diploid data frame) belongs to
population 1. Its genotype at <code>loc-1</code> is <code>44</code>,
thus homozygote for allele <code>4</code>. It is heterozygote for
alleles <code>3</code>and <code>4</code> at both <code>loc-2</code> and
<code>loc-3</code>, and homozygote for allele <code>3</code> at
<code>loc-4</code> and finaly homozygote for allele <code>4</code> at
<code>loc-5</code>. In fact, <code>loc-1</code> and <code>loc-4</code>
are monomorphic, meaning that only one allele is present in all
individuals from all populations.</p>
<p>If a genotype is missing, it is encoded as <code>NA</code>. For
instance, the fourth individual has not been typed at
<code>loc-3</code>, nor did the 6th individual for the same locus.</p>
<p>The first column of this dataframe contains the identifier of the
population to which the individual belongs. We can find how many
individuals were typed in each population by using the command
table:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(diploid[,<span class="dv">1</span>])</span></code></pre></div>
<pre><code>## 
## 1 2 3 4 5 6 
## 8 8 5 7 9 7</code></pre>
<p>As another example, we look at dataset <code>cont.isl99</code>, a
data frame where alleles are encoded as 2 digits numbers:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(cont.isl99)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cont.isl99)</span></code></pre></div>
<pre><code>##   Pop loc.1 loc.2 loc.3 loc.4 loc.5
## 1   1  7474  1955  9168  4051  9251
## 2   1  7474  3175  9168  2410  2327
## 3   1   808  3194  9536  9751  9223
## 4   1   874  5294  1876  1310  1292
## 5   1  7484  3875  1010  5107  7712
## 6   1   874  3175  1010  5135  9292</code></pre>
<p>the first individual is homozygous for allele <code>74</code> at the
first locus (<code>loc.1</code>) and heterozygous fore alleles
<code>19</code> and <code>55</code> at the second locus. The genotype
could have been written <code>5519</code> instead of <code>1955</code>,
it does not matter. Note the genotype of the 3rd and fourth individual
at the first locus. They both carry allele <code>8</code>, which is in
fact encoded as <code>08</code>. When it comes first, the leading 0
disappears, but it must be present in second position. Hence genotype
<code>874</code>, <code>0874</code> and <code>7408</code> are the same,
but different from genotype <code>748</code> who would be understood by
hierfstat as an individual heterozygous for alleles <code>07</code> and
<code>48</code>.</p>
<p>Last point: alleles for all loci to be analysed simultaneously must
be encoded with the same number of digits.</p>
</div>
<div id="importing-data-files" class="section level1">
<h1>Importing data files</h1>
<p>Often the data to be imported are in a text file. If this is the
case, the easiest way to import the file into <code>R</code> is via one
of the workhorse of R, the <code>read.table</code> function.</p>
<div id="importing-fstat-data-files" class="section level2">
<h2>Importing FSTAT data files</h2>
<p>If the data are in the <code>FSTAT</code> format, they can be readily
imported using the function <code>read.fstat</code>:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>dip<span class="ot">&lt;-</span>hierfstat<span class="sc">::</span><span class="fu">read.fstat</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,<span class="st">&quot;diploid.dat&quot;</span>,<span class="at">package=</span><span class="st">&quot;hierfstat&quot;</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dip)</span></code></pre></div>
<pre><code>##   Pop loc-1 loc-2 loc-3 loc-4 loc-5
## 1   1    44    43    43    33    44
## 2   1    44    44    43    33    44
## 3   1    44    44    43    43    44
## 4   1    44    44    NA    33    44
## 5   1    44    44    24    34    44
## 6   1    44    44    NA    43    44</code></pre>
</div>
<div id="importing-from-adegenet-genind-objects" class="section level2">
<h2>Importing from adegenet: genind objects</h2>
<p><code>adegenet</code> is another population genetics analysis
package, with the ability to import from several data format.
<code>hierfstat</code> can import and work directly with
<code>genind</code> objects generated by <code>adegenet</code>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(adegenet,<span class="at">quietly=</span><span class="cn">TRUE</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(nancycats)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">genind2hierfstat</span>(nancycats)[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]) <span class="co"># only the first 10 loci</span></span></code></pre></div>
<pre><code>##      pop   fca8  fca23  fca43  fca45  fca77  fca78  fca90  fca96  fca37
## N215 P01     NA 136146 139139 116120 156156 142148 199199 113113 208208
## N216 P01     NA 146146 139145 120126 156156 142148 185199 113113 208208
## N217 P01 135143 136146 141141 116116 152156 142142 197197 113113 210210
## N218 P01 133135 138138 139141 116126 150150 142148 199199  91105 208208
## N219 P01 133135 140146 141145 126126 152152 142148 193199 113113 208208
## N220 P01 135143 136146 145149 120126 150156 148148 193195  91113 208208</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(H3N2)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">genind2hierfstat</span>(H3N2,<span class="at">pop=</span><span class="fu">rep</span>(<span class="dv">1</span>,<span class="fu">dim</span>(H3N2<span class="sc">@</span>tab)[<span class="dv">1</span>]))[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]) <span class="co"># only the first 10 positions</span></span></code></pre></div>
<pre><code>##          pop X6 X17 X39 X42 X45 X51 X60 X72 X73
## AB434107   1  1   1   3   2   3   2   3   3   2
## AB434108   1  1   1   3   2   3   2   3   3   2
## AB438242   1 NA  NA  NA  NA  NA  NA   3   3   2
## AB438243   1 NA  NA  NA  NA  NA  NA   3   3   2
## AB438244   1 NA  NA  NA  NA  NA  NA   3   3   2
## AB438245   1 NA  NA  NA  NA  NA  NA   3   3   2</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(eHGDP)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">genind2hierfstat</span>(eHGDP))[,<span class="dv">1</span><span class="sc">:</span><span class="dv">11</span>] </span></code></pre></div>
<pre><code>##   pop  loc.1  loc.2  loc.3  loc.4  loc.5  loc.6  loc.7  loc.8  loc.9 loc.10
## 1   1 129155 264292 142146 156156 157166 171179 205205 183187 196196 137140
## 2   1 145150 288292 138142 168172 157166 171175 205210 195203 196196 128134
## 3   1 155155 292300 138142 156156 157169 167171 205205 183199 184196 137140
## 4   1 150155 264292 142146 156176 157163 171175 205205 187187 196196 140140
## 5   1 150155 292300 138146 156160 157166 171171 205205 183207 187190 128128
## 6   1 155155 296296 146146 152176 157157 167171 205210 179183 196196 128140</code></pre>
<p>The example below shows how to estimate gene diversities (expected
heterozygosities), observed heterozygosities, or estimate genetic
distances directly from a <code>genind</code> object:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">#basic.stats(nancycats)</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>hierfstat<span class="sc">::</span><span class="fu">Hs</span>(nancycats) <span class="co">#mean populations gene diversities</span></span></code></pre></div>
<pre><code>##       P01       P02       P03       P04       P05       P06       P07       P08 
## 0.6531333 0.7043000 0.7239111 0.7525222 0.6431222 0.7510000 0.6719333 0.7598556 
##       P09       P10       P11       P12       P13       P14       P15       P16 
## 0.6913667 0.7020111 0.7867667 0.6718333 0.6883889 0.7928556 0.7217111 0.7016222 
##       P17 
## 0.6046375</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>hierfstat<span class="sc">::</span><span class="fu">Ho</span>(nancycats) <span class="co"># mean populations observed heterozygosities</span></span></code></pre></div>
<pre><code>##       P01       P02       P03       P04       P05       P06       P07       P08 
## 0.5722222 0.5707111 0.6111111 0.6280222 0.5629667 0.6262667 0.5458556 0.6111111 
##       P09       P10       P11       P12       P13       P14       P15       P16 
## 0.7407556 0.6262667 0.6627444 0.5570000 0.6752222 0.6836667 0.7474889 0.6759222 
##       P17 
## 0.5980875</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">#genet.dist(nancycats)</span></span></code></pre></div>
</div>
</div>
<div id="allelic-dosages" class="section level1">
<h1>Allelic dosages</h1>
<p>Genomic datasets are composed of (very) large numbers of bi-allelic
loci, called Single Nucleotide Polymorhisms, or SNP. A convenient and
efficient format for storing SNP data is allelic dosage, the number of
alternative alleles and individual carries at a locus. For a diploid
locus, this number can be 0 if the individual carries two reference
alleles, 1 if he is heterozygote or 2 if he carries 2 alternate alleles.
The allelic dosage format is suitable for analyses with several of the
<code>hierfstat</code> functions, as described in the
<code>hierfstat</code> vignette.</p>
<div id="importing-from-adegenet-genlight-objects" class="section level2">
<h2>Importing from adegenet: genlight objects</h2>
<p>To import genlight objects from <code>adegenet</code>, just wrap the
object’s name in an <code>as.matrix</code>:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span><span class="fu">read.snp</span>(<span class="fu">system.file</span>(<span class="st">&quot;files/exampleSnpDat.snp&quot;</span>,<span class="at">package=</span><span class="st">&quot;adegenet&quot;</span>), <span class="at">chunk=</span><span class="dv">2</span>,<span class="at">quiet=</span><span class="cn">TRUE</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">as.matrix</span>(obj)</span></code></pre></div>
<pre><code>##                                                    1.a/t 8.g/c 11.a/c 43.t/a
## foo                                                    1     0      2      0
## bar                                                    0     0      1      2
## toto                                                   1     0     NA      0
## Nyarlathotep                                           0     1      2      0
## an even longer label but OK since on a single line     1     1      0      0</code></pre>
</div>
<div id="importing-vcf-files" class="section level2">
<h2>importing VCF files</h2>
<p><a href="https://samtools.github.io/hts-specs/">Variant Call Format
(VCF)</a> files have become a standard for genomic data. This is the
storage format used for the <a href="https://www.internationalgenome.org/">1000 genomes</a> for
instance.</p>
<ul>
<li>VCF files can be imported in <code>hierfstat</code> using the
function <code>read.VCF</code>, built from the <code>gaston</code>
package function <code>read.vcf</code>:</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gaston,<span class="at">quietly=</span><span class="cn">TRUE</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>filepath <span class="ot">&lt;-</span><span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;LCT.vcf.gz&quot;</span>, <span class="at">package=</span><span class="st">&quot;gaston&quot;</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>x1 <span class="ot">&lt;-</span> <span class="fu">read.VCF</span>( filepath )</span></code></pre></div>
<pre><code>## ped stats and snps stats have been set. 
## &#39;p&#39; has been set. 
## &#39;mu&#39; and &#39;sigma&#39; have been set.</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>x1</span></code></pre></div>
<pre><code>## A bed.matrix with 503 individuals and 607 markers.
## snps stats are set
## ped stats are set</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(x1)</span></code></pre></div>
<pre><code>## [1] 503 607</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(x1<span class="sc">@</span>snps,<span class="fu">table</span>(A1,A2))</span></code></pre></div>
<pre><code>##    A2
## A1    A   C   G   T
##   A   0  25  94  22
##   C  18   0  22 111
##   G 126  24   0  26
##   T  27  90  22   0</code></pre>
<p>by default, <code>read.VCF</code> keeps only the bi-allelic SNPs
(whereas <code>gaston::read.vcf</code> keeps all variants), but you can
choose to keep all variant by setting the argument
<code>BiAllelic</code> to FALSE.</p>
<p><code>as.matrix(x1)</code> will then provide allelic dosages.</p>
<ul>
<li>Another route to import VCF files is via the <code>SNPRelate</code>
package, and its function <code>snpgdsVCF2GDS</code> (as the function
writes to a file, the two next code chunks will not be evaluated)</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SNPRelate)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">snpgdsVCF2GDS</span>(filepath, <span class="st">&quot;test1.gds&quot;</span>, <span class="at">method=</span><span class="st">&quot;biallelic.only&quot;</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">snpgdsSummary</span>(<span class="st">&quot;test1.gds&quot;</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>f<span class="ot">&lt;-</span><span class="fu">snpgdsOpen</span>(<span class="st">&quot;test1.gds&quot;</span>)</span></code></pre></div>
<p><code>snpgdsVCF2GDS</code> stores the number of reference alleles, we
want the the number of alternate alleles:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>x2<span class="ot">&lt;-</span><span class="dv">2</span><span class="sc">-</span><span class="fu">snpgdsGetGeno</span>(f) </span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co">#check that allele frequencies are the same with the two methods</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(<span class="fu">colMeans</span>(x2)<span class="sc">/</span><span class="dv">2</span>,<span class="fu">colMeans</span>(<span class="fu">as.matrix</span>(x1)<span class="sc">/</span><span class="dv">2</span>),<span class="at">check.names=</span><span class="cn">FALSE</span>) </span></code></pre></div>
<ul>
<li>Last, we could import a <code>bed</code> file using
<code>gaston</code>, and use an external program, <a href="https://www.cog-genomics.org/plink/2.0/">plink</a>, to convert the
VCF file into a BED file.</li>
</ul>
<p>These are only 3 possibilities that I find convenient and efficient,
but many other exist, thus feel free to import dosage data into
<code>hierfstat</code> by your preferred route.</p>
</div>
</div>
<div id="data-generated-by-external-genetic-data-simulators" class="section level1">
<h1>Data generated by external genetic data simulators</h1>
<p>While <code>hierfstat</code> contains several built-in function to
generate genetic data (e.g. <code>sim.genot</code>), functions exist to
import data from external, more sophisticated genetic data simulators,
namely <a href="https://www2.unil.ch/popgen/softwares/quantinemo/">QuantiNemo</a>
and <a href="http://home.uchicago.edu/~rhudson1/source/mksamples.html">ms</a></p>
<div id="importing-from-quantinemo" class="section level2">
<h2>Importing from Quantinemo</h2>
<p><a href="https://www2.unil.ch/popgen/softwares/quantinemo/">QuantiNemo</a>
is a genetic simulation program for markers and traits. Data generated
by <code>Quantinemo</code> can be imported using the function
<code>read.fstat</code> if the <code>save_genotype</code> setting in
<code>quantinemo</code> is set to <span class="math inline">\(1\)</span>. If the <code>QuantiNemo</code> output
is set to <span class="math inline">\(2\)</span> (extended), 6 extra
columns are outputed and these can be read with
<code>hierfstat</code>using the function <code>qn2.read.fstat</code>.
The component <code>$dat</code> of the object return by this function
contains the genotypes of the individuals simulated, while the component
<code>$sex</code>contains its sex, the component <code>$ped</code> the
individuals’ pedigree and the component <code>$w</code> their fitness.
For more details about the extended <code>FSTAT</code> format of
<code>QuantiNemo</code>, see its manual.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>dat<span class="ot">&lt;-</span><span class="fu">qn2.read.fstat</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,<span class="st">&quot;qn2_sex.dat&quot;</span>,<span class="at">package=</span><span class="st">&quot;hierfstat&quot;</span>))</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(dat)</span></code></pre></div>
<pre><code>## [1] &quot;dat&quot; &quot;sex&quot; &quot;ped&quot; &quot;W&quot;</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dat<span class="sc">$</span>sex)</span></code></pre></div>
<pre><code>## [1] &quot;F&quot; &quot;F&quot; &quot;F&quot; &quot;F&quot; &quot;F&quot; &quot;F&quot;</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dat<span class="sc">$</span>dat[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code></pre></div>
<pre><code>##   Pop trait-1_locus-1 trait-1_locus-2 trait-1_locus-3 trait-1_locus-4
## 1   1             606            1515             101             404
## 2   1             606            1515             101             404
## 3   1             606            1515             101             404
## 4   1             606            1515             101             404
## 5   1             606            1515             101             404
## 6   1             606            1515             101             404
##   trait-1_locus-5 trait-1_locus-6 trait-1_locus-7 trait-1_locus-8
## 1             707             404             303             101
## 2             707             415             303             101
## 3             707             404             303             101
## 4             707             415             303             101
## 5             707             415             303             101
## 6             707            1504             303             101
##   trait-1_locus-9
## 1             808
## 2             808
## 3             808
## 4             808
## 5             808
## 6             808</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co">#sexbias.test(dat[[1]],sex=dat[[2]])</span></span></code></pre></div>
</div>
<div id="importing-from-ms" class="section level2">
<h2>Importing from ms</h2>
<p>The program <a href="http://home.uchicago.edu/~rhudson1/source/mksamples.html">ms</a>
of Hudson is commonly used to generate genomic data, and the program <a href="https://tskit.dev/msprime/docs/stable/intro.html">msprime</a> has
a wrapper for <code>ms</code> named <a href="https://tskit.dev/msprime/docs/stable/cli.html#sec-mspms">mspms</a>.</p>
<p>I very briefly discussed the output of the <code>ms</code> software
below, see its <a href="https://uchicago.app.box.com/s/l3e5uf13tikfjm7e1il1eujitlsjdx13/file/245765534272">documentation</a>
for more informations and details of arguments to be passed to
<code>ms</code>.</p>
<p>The output looks like this:</p>
<blockquote>
<p>ms 200 100 -t 20 -I 2 100 100 40 -n 2 0.01</p>
<p>29161</p>
<p>//</p>
<p>segsites: 23</p>
<p>positions: 0.0689 0.2534 0.3219 0.3350 0.3547 0.3768 0.4339 0.4359
0.4388 0.4694 0.5003 0.5422 0.6575 0.6985 0.7059 0.7147 0.7453 0.7709
0.7891 0.8439 0.8779 0.8857 0.9380</p>
<p>00100001100000000000000</p>
<p>00001001000000000001000</p>
</blockquote>
<p>The first line is the <code>ms</code> command line. It instructs the
program to simulate 2 populations (<code>-I 2</code>) , with <span class="math inline">\(\theta=2N_0\mu=20\)</span> (<code>-t 20</code>).
The 2 populations differ in size and the smallest (the second) is a
100th of the first (<code>-n 2 0.01</code>). The two populations
exchange <span class="math inline">\(4Nm=40\)</span> migrants per
generation and 100 chromosomes are sampled from each population
(<code>-I 2 100 100 40</code>), and this is repeated a 100 times (hence
<code>ms 200 100</code>).</p>
<p>The second line is a random number allowing to reapeat exactly the
same sequence</p>
<p>The third line is a line of comments</p>
<p>The fourth line contains the positions of the different segregating
sites</p>
<p>The fifth and following lines contain the genetic data, as a serie of
0 and 1 (0 is the ancestral and 1 the derived allele), collated one to
the other. These are the SNP sites, with 0 being the ancestral state and
1 the derived state.</p>
<p>the function <code>ms2bed</code> will convert ms output to a bed
object:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>bed<span class="ot">&lt;-</span><span class="fu">ms2bed</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,<span class="st">&quot;2pops_asspop.txt&quot;</span>,<span class="at">package=</span><span class="st">&quot;hierfstat&quot;</span>))</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">as.matrix</span>(bed[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])) <span class="co">#first 10 columns of bed matrix</span></span></code></pre></div>
<pre><code>##   M_1 M_2 M_3 M_4 M_5 M_6 M_7 M_8 M_9 M_10
## 1   0   0   1   0   1   0   0   2   1    0
## 2   1   0   0   0   1   0   0   1   0    0
## 3   0   0   0   0   0   0   0   2   0    0
## 4   0   0   0   0   2   0   0   2   0    0
## 5   0   0   0   0   0   0   0   2   0    0
## 6   0   0   0   0   1   0   0   2   0    0</code></pre>
<p>Take some time to explore the structure of the <code>bed</code> s4
object (<code>str(bed)</code>), it is very useful. the <code>@ped</code>
slot contains info relevant to the individuals (their names, family id
etc…), while the <code>@snps</code> slot contains info relevant to the
SNPs (their name, position, chromosome, etc…). The dosage matrix is
extracted with <code>as.matrix(bed)</code>. This can be used as argument
to all functions containing <code>dosage</code>in their names, such as
<code>beta.dosage, pi.dosage, theta.Watt.dosage, TajimaD.dosage</code>.
For instance, <code>fs.dosage</code> will produce estimates of
populations specific <span class="math inline">\(F_{IS}\)</span> and
<span class="math inline">\(F_{ST}\)</span>.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fs.dosage</span>(<span class="fu">as.matrix</span>(bed),<span class="at">pop=</span><span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>,<span class="at">each=</span><span class="dv">50</span>)) <span class="co"># population specific FSTs</span></span></code></pre></div>
<pre><code>##           1      2    All
## Fis -0.0046 0.0342 0.0148
## Fst -0.1616 0.6179 0.2282</code></pre>
<p>As a more interesting example, we can reuse the <code>eHGDP</code>
dataset we’ve encountered previously (see <code>?eHGDP</code>), and
after having converted it to dosage data with the <code>fstat2dos</code>
function (there is a total of <span class="math inline">\(8170\)</span>
alleles in the data set), we can look at inbreeding within populations
and population structure using <code>fs.dosage</code>:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>dat.HGDP<span class="ot">&lt;-</span><span class="fu">genind2hierfstat</span>(eHGDP)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>dos.HGDP<span class="ot">&lt;-</span><span class="fu">fstat2dos</span>(dat.HGDP[,<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>fs.HGDP<span class="ot">&lt;-</span><span class="fu">fs.dosage</span>(dos.HGDP,<span class="at">pop=</span>dat.HGDP[,<span class="dv">1</span>])</span></code></pre></div>
<p>We may be interested in finding which populations have either a large
<span class="math inline">\(F_{ST}^P&gt;0.15\)</span> or a low one <span class="math inline">\(F_{ST}^P &lt;0\)</span>:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>eHGDP<span class="sc">@</span>other<span class="sc">$</span>popInfo[<span class="fu">which</span>(fs.HGDP<span class="sc">$</span>Fs[<span class="dv">2</span>,]<span class="sc">&gt;</span><span class="fl">0.15</span>),]</span></code></pre></div>
<pre><code>##        Population  Region Label Latitude Longitude
## 23      Colombian AMERICA    23      3.0     -68.0
## 24      Karitiana AMERICA    24    -10.0     -63.0
## 26           Pima AMERICA    26     29.0    -108.0
## 54          Surui AMERICA    54    -11.0     -62.0
## 63         Guaymi AMERICA    63      8.5     -82.0
## 64        Cabecar AMERICA    64      9.5     -84.0
## 68           Ache AMERICA    68    -24.0     -56.0
## 69       Kaingang AMERICA    69    -24.0     -52.5
## 71           Kogi AMERICA    71     11.0     -74.0
## 75    TicunaArara AMERICA    75     -4.0     -70.0
## 76 TicunaTarapaca AMERICA    76     -4.0     -70.0
## 79        Arhuaco AMERICA    79     11.0     -73.8</code></pre>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>eHGDP<span class="sc">@</span>other<span class="sc">$</span>popInfo[<span class="fu">which</span>(fs.HGDP<span class="sc">$</span>Fs[<span class="dv">2</span>,]<span class="sc">&lt;</span><span class="fl">0.0</span>),]</span></code></pre></div>
<pre><code>##        Population Region Label Latitude Longitude
## 27     BantuKenya AFRICA    27     -3.0      37.0
## 28 BantuSouthWest AFRICA    28    -21.0      18.7
## 29 BantuSouthEast AFRICA    29    -28.4      27.6
## 30       Mandenka AFRICA    30     12.0     -12.0
## 31         Yoruba AFRICA    31      8.0       5.0
## 32     BiakaPygmy AFRICA    32      4.0      17.0
## 33     MbutiPygmy AFRICA    33      1.0      29.0
## 34            San AFRICA    34    -21.0      20.0</code></pre>
<p>and you’ll see that samples from AFRICA have low Population specific
<span class="math inline">\(F_{ST}\)</span>, while samples from AMERICA
have large population specific <span class="math inline">\(F_{ST}\)</span></p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
